In [13]:
from tensorflow import keras
import numpy as np

In [2]:
input_layer = keras.layers.Input(shape=(150, 150, 3))
convolutional = keras.layers.Conv2D(filters=32, kernel_size=(3, 3), activation='relu')(input_layer)
max_pooling = keras.layers.MaxPool2D(pool_size=(2, 2))(convolutional)
flatten = keras.layers.Flatten()(max_pooling)
inner_dense = keras.layers.Dense(64, activation='relu')(flatten)
output_layer = keras.layers.Dense(1, activation='sigmoid')(inner_dense)

model = keras.Model(input_layer, output_layer)

In [3]:
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 150, 150, 3)]     0         
                                                                 
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2  (None, 74, 74, 32)        0         
 D)                                                              
                                                                 
 flatten (Flatten)           (None, 175232)            0         
                                                                 
 dense (Dense)               (None, 64)                11214912  
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                             

In [4]:
optimizer = keras.optimizers.SGD(lr=0.002, momentum=0.8)
loss_function = keras.losses.BinaryCrossentropy()
metrics = ['accuracy']

model.compile(optimizer=optimizer, loss=loss_function, metrics=metrics)

In [5]:
from keras.preprocessing.image import ImageDataGenerator

data_gen = ImageDataGenerator(rescale=1. / 255)
train_data = data_gen.flow_from_directory('./data/train', target_size=(150, 150), batch_size=20, shuffle=True, class_mode='binary')
test_data = data_gen.flow_from_directory('./data/test', target_size=(150, 150), batch_size=20, shuffle=True, class_mode='binary')

Found 3677 images belonging to 2 classes.
Found 918 images belonging to 2 classes.


we have 3677 images, each image size is 150*150 therefore the shape of one image is (150, 150, 3)

In [6]:
train_data.image_shape

(150, 150, 3)

batch size is 20 therefore we will have 3677/20 = 184 batches

In [7]:
len(train_data)

184

each batch consists of image (x) and its label (y)
therefore it will have two data in it, x and y

In [8]:
len(train_data[0])

2

In [9]:
history = model.fit(train_data, epochs=10, validation_data=test_data)

Epoch 1/10
184/184 [==============================] - 9s 48ms/step - loss: 0.6606 - accuracy: 0.5980 - val_loss: 0.6204 - val_accuracy: 0.6558
Epoch 2/10
184/184 [==============================] - 9s 48ms/step - loss: 0.6075 - accuracy: 0.6717 - val_loss: 0.6659 - val_accuracy: 0.5479
Epoch 3/10
184/184 [==============================] - 9s 47ms/step - loss: 0.5374 - accuracy: 0.7354 - val_loss: 0.5243 - val_accuracy: 0.7527
Epoch 4/10
184/184 [==============================] - 9s 46ms/step - loss: 0.4797 - accuracy: 0.7751 - val_loss: 0.5483 - val_accuracy: 0.7462
Epoch 5/10
184/184 [==============================] - 8s 46ms/step - loss: 0.4125 - accuracy: 0.8208 - val_loss: 0.5214 - val_accuracy: 0.7734
Epoch 6/10
184/184 [==============================] - 8s 45ms/step - loss: 0.3271 - accuracy: 0.8692 - val_loss: 0.5713 - val_accuracy: 0.7353
Epoch 7/10
184/184 [==============================] - 8s 46ms/step - loss: 0.2145 - accuracy: 0.9228 - val_loss: 0.6521 - val_accuracy: 0.7505

In [21]:
np.median(history.history['accuracy'])

0.9986401796340942

In [15]:
np.std(history.history['loss'])

0.2047529459950159

In [16]:
train_data_gen = ImageDataGenerator(
    rescale=1. / 255,
    rotation_range=50,
    width_shift_range=0.1,
    height_shift_range=0.1,
    zoom_range=0.1,
    horizontal_flip=True,
    fill_mode='nearest'
)

new_train_data = data_gen.flow_from_directory('./data/train', target_size=(150, 150), batch_size=20, shuffle=True, class_mode='binary')

Found 3677 images belonging to 2 classes.


In [17]:
history = model.fit(new_train_data, epochs=10, validation_data=test_data)

Epoch 1/10
184/184 [==============================] - 9s 46ms/step - loss: 0.0417 - accuracy: 0.9921 - val_loss: 0.9610 - val_accuracy: 0.7168
Epoch 2/10
184/184 [==============================] - 8s 45ms/step - loss: 0.0596 - accuracy: 0.9872 - val_loss: 1.3478 - val_accuracy: 0.6830
Epoch 3/10
184/184 [==============================] - 8s 45ms/step - loss: 0.0377 - accuracy: 0.9932 - val_loss: 1.1676 - val_accuracy: 0.7200
Epoch 4/10
184/184 [==============================] - 9s 47ms/step - loss: 0.0240 - accuracy: 0.9986 - val_loss: 1.2111 - val_accuracy: 0.7364
Epoch 5/10
184/184 [==============================] - 9s 47ms/step - loss: 0.0139 - accuracy: 0.9986 - val_loss: 1.0577 - val_accuracy: 0.7320
Epoch 6/10
184/184 [==============================] - 8s 44ms/step - loss: 0.0186 - accuracy: 0.9986 - val_loss: 1.2842 - val_accuracy: 0.7200
Epoch 7/10
184/184 [==============================] - 8s 45ms/step - loss: 0.0191 - accuracy: 0.9986 - val_loss: 1.2673 - val_accuracy: 0.7440

In [20]:
np.mean(history.history['val_loss'])

1.19269802570343

In [25]:
np.average(history.history['val_accuracy'][5:])

0.7355119824409485